clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -421,6 +421,230 @@ test_parallel()
+    }
+    scheduler_t scheduler;
+    if (scheduler.init(sched_inputs, NUM_OUTPUTS, std::move(sched_ops)) !=
+        scheduler_t::STATUS_SUCCESS)
+        assert(false);
+    std::unordered_map<memref_tid_t, int> tid2stream;
+    for (int i = 0; i < NUM_OUTPUTS; i++) {
+        auto *stream = scheduler.get_stream(i);
+        memref_t memref;
+
+        // Whether we're currently in the syscall sequence records.
+        bool in_syscall = false;
+        // Whether we just saw the syscall trace end, and still need to take into
+        // account the read-ahead record count.
+        bool saw_syscall_trace = false;
+        // The scheduler will buffer some read-ahead entries to determine the target
+        // for the indirect branch at the end of the syscall sequence. So we expect
+        // the input stream ordinals to run ahead by the following amount.
+        int expected_record_readahead = 0;
+        int expected_instr_readahead = 0;
+        uint64_t last_input_record_ordinal = 0;
+        uint64_t last_input_instr_ordinal = 0;
+        for (scheduler_t::stream_status_t status = stream->next_record(memref);
+             status != scheduler_t::STATUS_EOF; status = stream->next_record(memref)) {
+            assert(status == scheduler_t::STATUS_OK);
+            // Ensure one input thread is only in one output stream.
+            if (tid2stream.find(memref.instr.tid) == tid2stream.end()) {
+                tid2stream[memref.instr.tid] = i;
+            } else
+                assert(tid2stream[memref.instr.tid] == i);
+            bool at_syscall_start = false;
+            bool at_syscall_end = false;
+            if (memref.marker.type == TRACE_TYPE_MARKER) {
+                if (memref.marker.marker_type == TRACE_MARKER_TYPE_SYSCALL_TRACE_START) {
+                    at_syscall_start = in_syscall = true;
+                } else if (memref.marker.marker_type ==
+                           TRACE_MARKER_TYPE_SYSCALL_TRACE_END) {
+                    at_syscall_end = true;
+                }
+            }
+            if (at_syscall_start) {
+                // We read-ahead by one record to determine the syscall injection
+                // point.
+                expected_record_readahead = 1;
+            } else if (in_syscall && type_is_instr_branch(memref.instr.type) &&
+                       !type_is_instr_direct_branch(memref.instr.type)) {
+                // We read-ahead until the next instr after the indirect branch
+                // instr at the end of the syscall trace.
+                expected_record_readahead = 5;
+                expected_instr_readahead = 1;
+            } else if (at_syscall_end) {
+                // From next-iteration we begin counting down the expected
+                // read-ahead records.
+                saw_syscall_trace = true;
+            } else if (saw_syscall_trace) {
+                // The record count considered read ahead of their time will reduce
```

Comment:
Having trouble understanding this sentence: could it be rewritten to clarify? "read" present vs past tense with "readahead" priming for present tense is part of the problem. "to be read" would help.

Reply:
Done


                // The record count considered to be read from the input stream
                // ahead of their time will reduce by one each iteration, starting
                // from the record after the syscall_trace_end marker.